torch/optim/nadam.py

```diff
@@ -242,10 +248,21 @@ def _single_tensor_nadam(params: List[Tensor],
         # decay the first and second moment running average coefficient
         exp_avg.mul_(beta1).add_(grad, alpha=1 - beta1)
         exp_avg_sq.mul_(beta2).addcmul_(grad, grad, value=1 - beta2)
-
-        denom = exp_avg_sq.div(bias_correction2).sqrt().add_(eps)
-        param.addcdiv_(grad, denom, value=-lr * (1. - mu) / (1. - mu_product.item()))
-        param.addcdiv_(exp_avg, denom, value=-lr * mu_next / (1. - mu_product_next.item()))
+        denom = exp_avg_sq.div(bias_correction2).sqrt()
+        if differentiable:
+            denom = denom.add(eps)
+            # Make autograd track the operations
+            grad = grad * (-lr * (1. - mu) / (1. - mu_product))
+            exp_avg = grad * (-lr * (1. - mu_next) / (1. - mu_product_next))
+            value_grad = 1.0
+            value_exp_avg = 1.0
+        else:
+            denom.add_(eps)
+            value_grad = (-lr * (1. - mu) / (1. - mu_product)).item()
```

Comment:
is item needed here in fact?

Reply:
yeah .. `addcdiv_` requires `value` to be a python scalar. I guess we could unify both execution paths to avoid device synchronization but not sure. cc @albanD (who also was concerned about this)